TOP-9 feat: Signup Data모듈 작성

cwj-c opened the pull request:
## Explain this Pull Request 🙏

- Signup Data모듈 작성

## What has changed? 🤔
- 변경 사항 
  - Data모듈에 Signup Entity 작성, Mapper 함수 작성
  - Data모듈 assets에 회원가입 이용약관 json 파일 추가
  - Data모듈에 Signup Dao 작성
  - Data모듈에 Signup DataSource 작성
  - Data모듈에 Signup Repository 구현체 작성
  - Data, Domain모듈의 회원가입 클래스들의 di 코드 작성
  - Data모듈의 Local 쪽 DataSource, Repository Method 테스트 케이스 작성

## Screenshot 📸


## Test Checklist 🚩

cwj-c committed:
modify: TOP-9 TermsModel 수정

 - TermsModel.TermsContent 정의
 - TermsModel의 content를 List<TermsContent>로 변경
 - 변경에 따른 테스트 케이스 수정

cwj-c committed:
feat: TOP-9 회원가입 이용약관 json file을 assets에 추가

cwj-c committed:
feat: TOP-9 THTApiConstant 작성, 적용

cwj-c committed:
feat: TOP-9 TermsEntity 정의 및 Mapper 함수 작성

cwj-c committed:
feat: TOP-9 SignupResponse 정의

 - 임시로 정의된 클래스

cwj-c committed:
feat: TOP-9 THTSignupApi 정의, 네이밍 변경

 - 회원가입 API Service 정의

cwj-c committed:
rename: TOP-9 기존 Api네이밍의 의미를 세분화 하도록 변경

cwj-c committed:
feat: TOP-9 TermsDao, 구현체 작성

 - android assets resource에서 terms를 불러옴

cwj-c committed:
feat: TOP-9 TermsDataSource, 구현체 작성

cwj-c committed:
feat: TOP-9 SignupUserEntity 정의 및 Mapper 함수 추가

cwj-c committed:
feat: TOP-9 SignupUserDataSource, 구현체 작성

 - SharedPrererneces를 주입받아 회원가입 프로세스를 진행중인 유저 데이터를 local에서 관리

cwj-c committed:
feat: TOP-9 THTSignupApi 리턴 결과 타입을 ThtResponse로 수정

cwj-c committed:
feat: TOP-9 IdealTypeModel, InterestModel에 emojiCode 추가 및 Mapper 함수 작성

cwj-c committed:
feat: TOP-9 OkHttpInterceptorModule의 Interceptor 타입의 provide 메소드가 2개인 문제 수정

cwj-c committed:
feat: TOP-9 SignupApiDataSource, 구현체 작성

cwj-c committed:
feat: TOP-9 SignupRepositoryImpl 작성, 테스트 케이스 작성

cwj-c committed:
remove: TOP-9 사용하지 않는 클래스, 파일 제거

cwj-c committed:
test: TOP-9 TermsDataSource 테스트 케이스 작성

cwj-c committed:
modify: TOP-9 SignupUserDataSourceImpl의 signupUserMap이 첫 시도에 없을 시 생성해주는 코드 추가

cwj-c committed:
test: TOP-9 coroutine-test, mockk, assertj android test 의존성 추가

cwj-c committed:
test: TOP-9 SignupUserDatSource 테스트 케이스 작성

cwj-c committed:
remove: TOP-9 사용하지 않는 테스트 케이스 제거

cwj-c committed:
rename: TOP-9 PatchSignupBirthUseCase 오타 수정

cwj-c committed:
fea: TOP-9 App모듈에 AppModule 작성

cwj-c committed:
fea: TOP-9 Data모듈에 DaoModule 작성

cwj-c committed:
fea: TOP-9 Data모듈에 DataSourceModule 작성

cwj-c committed:
fea: TOP-9 Data모듈에 DispatcherModule 작성

cwj-c committed:
fea: TOP-9 Data모듈에 RepositoryModule 작성

cwj-c committed:
fea: TOP-9 Data모듈에 SharedPreferencesModule 작성

cwj-c committed:
fea: TOP-9 Data모듈에 UseCaseModule 작성

cwj-c committed:
feat: TOP-9 THTSignupApi의 일부 api에 HTTP Method 종류를 기입하지 않은 문제 수정

cwj-c committed:
refactor: TOP-9 ApiDataSource를 remote/datasource 패키지로 이동

cwj-c committed:
refactor: TOP-9 local DataSource들을 local/datasource 패키지로 이동

taewooyo commented on data/src/main/java/com/tht/tht/data/di/retrofit/ApiServiceModule.kt:
```diff
@@ -10,11 +10,11 @@ import javax.inject.Singleton
 @Module
 @InstallIn(SingletonComponent::class)
-internal class ApiModule {
+object ApiServiceModule {
```
Object로 변경을 한 의도가 궁금합니다!

cwj-c replied:
>In Kotlin, modules that only contain @Provides functions can be object classes. This way, providers get optimized and almost in-lined in generated code.

@Provides만 사용하는 모듈의 경우 object로 설정하면 최적화되고, 생성된 코드가 인라이닝 된다고 [hilt-codelab 문서](https://developer.android.com/codelabs/android-hilt#6)에 언급이 되어있어서 object로 변경하였습니다!!

taewooyo replied:
오 감사합니다!

taewooyo commented on data/src/main/java/com/tht/tht/data/local/dao/TermsDaoImpl.kt:
```diff
@@ -0,0 +1,23 @@
+        return try {
+            val inputStream = context.resources.assets.open("signup_terms.json")
+            val reader = inputStream.bufferedReader()
+            Gson().fromJson(reader, TermsEntity::class.java)
```
DI에 GsonModule이 있는데 이용해보는 건 어떤가요?

cwj-c replied:
Gson을 사용하는 코드가 몇줄 안되다보니 미처 인지하지 못하였네요!!
적용해서 사용해보도록 하겠습니다!

cwj-c committed:
feat: TOP-9 SignupUserDao, 구현체 작성

 - SignupUser를 Local에 직접적으로 관리하는 역할과, Local 혹은 Remote에서 SignupUser를 호출하는 책임을 분리하기위해 SignupUserDao를 정의, SignupUser를 Local에서 관리하는 코드를 이동시킴

cwj-c committed:
feat: TOP-9 SignupUserDao di 코드 작성

cwj-c committed:
test: TOP-9 SignupUserDao 테스트 케이스 작성

cwj-c committed:
modify: TOP-9 SignupUserDataSource 구현체 수정

 - SignupUser를 Local에 직접적으로 관리하는 역할과, Local 혹은 Remote에서 SignupUser를 호출하는 책임을 분리하기위해 SignupUserDao를 정의, SignupUser를 Local에서 관리하는 코드를 이동시킴
 - SignupUserDataSourceImpl에서는 Local에서 SignupUser를 관리하는 SignupUserDao를 올바른 Dispatcher에서 호출하여 결과를 리턴하는 역할을 수행

cwj-c committed:
test: TOP-9 변경된 SignupUserDataSource 테스트 케이스 작성

cwj-c committed:
remove: TOP-9 변경되기 이전 SignupUserDataSource 테스트 케이스 제거

cwj-c commented:
SignupUserDataSourceImpl에서 Local(SharedPreferences)에 SignupUser를 저장, 삭제, 불러오기 등 구체적인 관리 코드와, SignupUserDataSource의 구현 코드가 함께 존재하고 있었습니다.

SignupUser를 Local에 관리하는 코드는 DaoImpl로 분리하고, 
DataSource에서는 Dao의 메소드를 호출하고, 그 결과를 판단하여 리턴하는 역할을 수행하게끔 하여 관심사를 분리하는게 
좋다고 판단하여 분리 작업, 테스트 케이스 수정 작업을 진행하였습니다

Room을 사용할때 구체적인 Dao Method(SQL Query)들의 구현은 Room에서 제공해주는것과 유사하게, Dao구현체에서 직접적인 Local 처리 코드를 포함하는게 좋다고 판단했습니다

kwangddang commented on data/src/main/java/com/tht/tht/data/di/DaoModule.kt:
```diff
@@ -0,0 +1,21 @@
+abstract class DaoModule {
+
+    @Binds
+    abstract fun bindTermsDao(impl: TermsDaoImpl): TermsDao
```
dao같은 경우 @Singleton 키워드를 붙여서 만들어야 하지 않나요??

cwj-c replied:
`SignupUserDaoImpl`의 경우 `SharedPreferences`를 주입받아 사용을 하는데, `SharedPreferences`의 경우 생성자의 String값을 통해 공유 파일을 식별하기에, 해당 String값이 같다면 동일한 SharedPreferences Data에 접근한다고 할 수 있습니다.
때문에 Singleton으로 지정하지 않았습니다

`TermsDaoImpl`의 경우는 `fetchTerms()` 메소드가 호출될때 마다 asset에서 파일을 불러와 리턴하기 때문에 Singleton으로 설정할 필요성을 찾지 못했습니다.

이러한 이유들로 인해 Singleton으로 설정하지 않았는데.. 혹시 제 생각이 잘못되었거나 Singleton으로 지정되어야 할 다른 이유가 있을까요??

kwangddang replied:
Dao를 Singleton으로 지정하면 다음과 같은 장점이 있다고 하네요!
- DAO 인스턴스를 공유함으로써 데이터베이스 연결을 최소화할 수 있음. 
  - 데이터베이스 연결은 비용이 많이 드는 작업 중 하나이기 때문에, 인스턴스를 공유함으로써 이 비용을 줄일 수 있음.
- DAO 인스턴스가 유지하는 상태 정보를 계속 유지할 수 있기 때문에, DAO 메서드를 호출할 때마다 동일한 상태를 사용할 수 있습니다.
  - (크게 와닿지는 않는 것 같습니다..)

여기까지 써놓고 조금 더 찾아보니
애초에 `DataBase`를 생성할 때 Singleton으로 만들면 자동으로 `Dao`도 Singleton이 되어서
`@Singleton` 어노테이션이 없어도 Singleton이 된다고 하네요.
그러니 바꾸지 않아도 될 것 같습니다!

(추가로 기능상의 차이는 없지만 코드를 명확하게 하기 위해 `@Singleton` 어노테이션을 사용하여 DAO 클래스가 명시적으로 Singleton임을 나타내는 경우도 있다는 것 같습니다 ㅎㅎ)

cwj-c replied:
Dao가 Singleton으로 설정되었을때 장점에 대해 이해된것 같습니다!!
광현님 답변을 보니, 현재 Dao구현체가 DB에 접근하지는 않지만,  굳이 Singleton으로 하지 않을 이유가 없을 것 같다는 생각이 들었습니다.
또한, 코드를 명확하게 하는 용도로라도 `@Singleton`어노테이션을 사용하는게 좋을 것 같다는 생각이 들어 코드에 적용하도록 하겠습니다!!ㅎㅎ

kwangddang commented on data/src/main/java/com/tht/tht/data/local/dao/SignupUserDaoImpl.kt:
```diff
@@ -0,0 +1,44 @@
+    private var signupUserMap: HashMap<String, SignupUserEntity>
+        get() {
+            val typeToken = object : TypeToken<HashMap<String, SignupUserEntity>>() {}.type
+            return gson.fromJson(sp.getString(SIGNUP_USER_KEY, ""), typeToken) ?: HashMap()
```
TypeToken에 대해 공부해보고 다시 리뷰를 하러 오겠습니다..

cwj-c replied:
이 [링크](https://namocom.tistory.com/671)를 확인해보시면 도움이 될 것 같습니다!!

제가 요약해서 설명을 드리자면, 
JVM의 Generic의 경우 타입소거를 사용해서 구현되기에 런타임에 타입에 대한 정보를 알 수 없습니다
때문에 List<T>와 같은 타입을 Deserialize할때, 어떤 타입으로 변환해야하는지 알 수 없습니다
(Serialize할때는 Array형태로 변환하기에, 특별한 규칙이 없어도 가능하다고 합니다)

따라서 런타임에 Deserialize를 수행할때, Generic Type의 표현을 제공해줘야 하는데 TypeToken클래스가 그 역할을 수행해준다고 합니다

kwangddang replied:
완벽히 이해한 건 아니지만 어느정도 감은 잡은 것 같습니다 감사합니다!

cwj-c committed:
feat: TOP-9 SignpUserDaoImpl에 Singleton annotation 적용

cwj-c committed:
feat: TOP-9 TermsDaoImpl에 Singleton annotation 적용
